### PR TITLE
improve di performance for cache

### PR DESCRIPTION
**Pull request by icewind1991**
not really the cleanest, but `Cache` is to deeply integrated to add a ton of DI parameters without being a major pain

### CODE REVIEW

**Review by AndyScherzinger — APPROVED** [review]

File: apps/files_external/tests/Service/StoragesServiceTest.php [hunk 1]
```diff
@@ -315,7 +316,7 @@ public function testDeleteStorage($backendOptions, $rustyStorageId) {
 
 		// manually trigger storage entry because normally it happens on first
 		// access, which isn't possible within this test
-		$storageCache = new \OC\Files\Cache\Storage($rustyStorageId);
+		$storageCache = new \OC\Files\Cache\Storage($rustyStorageId, true, \OC::$server->get(IDBConnection::class));
```
Comment (skjnldsv) [inline apps/files_external/tests/Service/StoragesServiceTest.php:319]:
```suggestion
		use OCP\Server;
		$storageCache = new \OC\Files\Cache\Storage($rustyStorageId, true, Server::get(IDBConnection::class));
```

File: apps/files_external/tests/Service/StoragesServiceTest.php [hunk 2]
```diff
@@ -44,6 +44,7 @@
 use OCP\Files\Config\IUserMountCache;
 use OCP\Files\Mount\IMountPoint;
 use OCP\Files\Storage\IStorage;
+use OCP\IDBConnection;
 use OCP\IUser;
```
Comment (skjnldsv) [inline apps/files_external/tests/Service/StoragesServiceTest.php:48]:
```suggestion
use OCP\IDBConnection;
use OCP\IUser;
use OCP\Server;
```

File: apps/files_external/tests/Service/StoragesServiceTest.php (hunk 1 above)
Comment (skjnldsv) [inline apps/files_external/tests/Service/StoragesServiceTest.php:319]:
```suggestion
		$storageCache = new \OC\Files\Cache\Storage($rustyStorageId, true, Sserver::get(IDBConnection::class));
```

File: lib/private/Files/Cache/Wrapper/CacheWrapper.php [hunk 3]
```diff
@@ -30,33 +30,31 @@
...
+		if (!$dependencies && $cache instanceof Cache) {
 			$this->mimetypeLoader = $cache->mimetypeLoader;
 			$this->connection = $cache->connection;
 			$this->querySearchHelper = $cache->querySearchHelper;
 		} else {
-			$this->mimetypeLoader = \OC::$server->get(IMimeTypeLoader::class);
-			$this->connection = \OC::$server->get(IDBConnection::class);
-			$this->querySearchHelper = \OC::$server->get(QuerySearchHelper::class);
+			if (!$dependencies) {
+				$dependencies = \OC::$server->get(CacheDependencies::class);
```
Comment (skjnldsv) [inline lib/private/Files/Cache/Wrapper/CacheWrapper.php:53]:
```suggestion
				$dependencies = Server::get(CacheDependencies::class);
```

File: lib/private/Files/Cache/Wrapper/CacheWrapper.php [hunk 4]
```diff
@@ -30,33 +30,31 @@
 namespace OC\Files\Cache\Wrapper;
 
 use OC\Files\Cache\Cache;
-use OC\Files\Cache\QuerySearchHelper;
+use OC\Files\Cache\CacheDependencies;
 use OCP\Files\Cache\ICache;
 use OCP\Files\Cache\ICacheEntry;
-use OCP\Files\IMimeTypeLoader;
 use OCP\Files\Search\ISearchOperator;
 use OCP\Files\Search\ISearchQuery;
```
Comment (skjnldsv) [inline lib/private/Files/Cache/Wrapper/CacheWrapper.php:37]:
```suggestion
use OCP\Files\Search\ISearchQuery;
use OCP\Server;
```

File: lib/private/Files/Storage/Common.php [hunk 5]
```diff
@@ -338,12 +339,20 @@ public function hasUpdated($path, $time) {
 		return $this->filemtime($path) > $time;
 	}
 
+	protected function getCacheDependencies(): CacheDependencies {
+		static $dependencies = null;
+		if (!$dependencies) {
+			$dependencies = \OC::$server->get(CacheDependencies::class);
```
Comment (skjnldsv) [inline lib/private/Files/Storage/Common.php:345]:
```suggestion
			$dependencies = Server::get(CacheDependencies::class);
```

**Review by skjnldsv — CHANGES_REQUESTED** [review]
See comments

**Comment by icewind1991** [comment]
> See comments

fixed

**Review by skjnldsv — APPROVED** [review]

**Pull request merged**